### PR TITLE
Verify classpath elements

### PR DESCRIPTION
**Pull request by nre-ableton**
This will help to safeguard against invalid versions passed to the
script.

---

ping @AbletonDevTools/gotham-city

### CODE REVIEW

File: run_codenarc.py [hunk 1]
```diff
@@ -28,6 +28,26 @@ def __init__(self, num_violations):
         self.num_violations = num_violations
 
 
+def _build_classpath(args):
+    """Construct a classpath from the Groovylint home directory and JAR versions."""
```
Comment (ala-ableton) [inline run_codenarc.py]: `Groovylint` -> `Groovy`

Comment (ala-ableton) [reply run_codenarc.py]: Actually, it's both :smile:

File: run_codenarc.py [hunk 2]
```diff
@@ -45,6 +45,10 @@ def _build_classpath(args):
         f'{home_dir}/slf4j-{slf4j_version}/slf4j-simple-{slf4j_version}.jar',
     ]
 
+    for path in classpath:
+        if not os.path.exists(path) and not path.endswith('*'):
```
Comment (ala-ableton) [inline run_codenarc.py]:
```python
        if not(os.path.exists(path) or path.endswith('*')):
```
might be easier to understand.

File: tests/test_run_codenarc.py [hunk 3]
```diff
@@ -108,6 +110,21 @@ def test_run_codenarc_failure_code():
...
 
+def test_run_codenarc_missing_jar():
+    """Test that run_codenarc raises an error if a JAR file could not be found.
+
+    Calling run_codenarc with no valid version arguments should result in no files found,
+    which should cause _build_classpath to raise.
+    """
+    with pytest.raises(ValueError) as exception_info:
+        run_codenarc(
+            args=parse_args(args=['--codenarc-version', '6.6.6']),
```
Comment (ala-ableton) [inline tests/test_run_codenarc.py]: You don't have to pass `--codenarc-version 6.6.6`, you will still get a `ValueError` exception raised.

Comment (nre-ableton) [reply tests/test_run_codenarc.py]: Yes, but I want to make sure that the exception raised is for this specific case and not some other random `ValueError`.

File: run_codenarc.py (hunk 1 above)
Comment (nre-ableton) [reply run_codenarc.py]: Fixed in 1cafaae.

File: run_codenarc.py (hunk 2 above)
Comment (nre-ableton) [reply run_codenarc.py]: Fixed in 1bd3ebd.

File: run_codenarc.py [hunk 4]
```diff
@@ -46,7 +46,7 @@ def _build_classpath(args):
     ]
 
     for path in classpath:
-        if not os.path.exists(path) and not path.endswith('*'):
+        if not(os.path.exists(path) or path.endswith('*')):
```
Comment (ala-ableton) [inline run_codenarc.py]: I would expect a space after the `not`, but `flake8` doesn't seem to complain... :thinking:

Comment (nre-ableton) [reply run_codenarc.py]: IntelliJ formatted it that way. 🤷‍♂️ I'll add a space, though.

**Comment by nre-ableton** [comment]
@ala-ableton Done!

File: run_codenarc.py [hunk 5]
```diff
@@ -28,6 +28,26 @@ def __init__(self, num_violations):
         self.num_violations = num_violations
 
 
+def _build_classpath(args):
+    """Construct the classpath from the Groovy/Groovylint homes and JAR versions."""
+    home_dir = args.home
+    groovy_home = args.groovy_home
+    codenarc_version = args.codenarc_version
+    gmetrics_version = args.gmetrics_version
+    slf4j_version = args.slf4j_version
```
Comment (ala-ableton) [inline run_codenarc.py]: All these variables could be inlined:
```python
def _build_classpath(args):
    """Construct the classpath from the Groovy/Groovylint homes and JAR versions."""
    classpath = [
        args.home,
        f'{args.groovy_home}/lib/*',
        f'{args.home}/CodeNarc-{args.codenarc_version}.jar',
        f'{args.home}/GMetrics-{args.gmetrics_version}.jar',
        f'{args.home}/slf4j-{args.slf4j_version}/slf4j-api-{args.slf4j_version}.jar',
        f'{args.home}/slf4j-{args.slf4j_version}/slf4j-simple-{args.slf4j_version}.jar',
    ]

    return ':'.join(classpath)
```

Comment (nre-ableton) [reply run_codenarc.py]: True, fixed in a9167f2.

**Pull request merged**